src/fl/federation/client.py

```diff
@@ -47,32 +47,32 @@ def fit(self, parameters, config):
 
-    def calculate_r2_score(self, trainer: Trainer, loader: DataLoader) -> float:
+    def calculate_loader_metrics(self, trainer: Trainer, loader: DataLoader) -> Tuple[float, float]:
         preds = trainer.predict(self.model, loader)
-        preds = torch.cat(preds, dim=0)
-        y_true = torch.cat([batch[1] for batch in iter(loader)])
-        r2 = r2_score(y_true.detach().cpu().numpy(), preds.detach().cpu().numpy())
-        return r2
-
-    def calculate_train_val_r2(self, trainer: Trainer) -> float:
-        train_loader, val_loader = self.data_module.predict_dataloader()
+        preds = torch.cat(preds, dim=0).detach().cpu().numpy()
+        y_true = torch.cat([batch[1] for batch in iter(loader)]).detach().cpu().numpy()
         
-        return self.calculate_r2_score(trainer, train_loader), self.calculate_r2_score(trainer, val_loader)
+        mse = mean_squared_error(y_true, preds)
+        r2 = r2_score(y_true, preds)
+        # we do that because mse and r2 have type numpy.float32 which is not a valid type for return of `evaluate` function
+        return float(mse), float(r2)
 
     def evaluate(self, parameters, config):
         self.set_parameters(parameters)
 
         trainer = Trainer(logger=self.logger, **self.training_params)
         # train_loader = DataLoader(self.model.train_dataset, batch_size=64, num_workers=1, shuffle=False)
         
-        val_results = trainer.validate(self.model, datamodule=self.data_module, verbose=False)
-        val_loss = val_results[0]["val_loss"]
-        train_r2, val_r2 = self.calculate_train_val_r2(trainer)
+        train_loader, val_loader = self.data_module.predict_dataloader()
+        train_loss, train_r2 = self.calculate_loader_metrics(trainer, train_loader)
+        val_loss, val_r2 = self.calculate_loader_metrics(trainer, val_loader)
 
         val_len = self.data_module.val_len()
-        round = self.model.current_round
-        logging.info(f'round: {round}\ttrain_r2: {train_r2:.4f}\tval_r2: {val_r2:.4f}\tval_loss: {val_loss:.3f}\tval_len: {val_len}')
-        mlflow.log_metric('train_r2', train_r2, round)
-        mlflow.log_metric('val_r2', val_r2, round)
-        mlflow.log_metric('val_loss', val_loss, round)
-        return val_loss, val_len, {"val_loss": val_loss, "train_r2": train_r2, "val_r2": val_r2}
+        epoch = self.model.fl_current_epoch()
+        logging.info(f'round: {self.model.current_round}\ttrain_loss: {train_loss:.4f}\ttrain_r2: {train_r2:.4f}\tval_loss: {val_loss:.3f}\tval_r2: {val_r2:.4f}\tval_len: {val_len}')
+        mlflow.log_metric('train_loss', train_loss, epoch)
+        mlflow.log_metric('train_r2', train_r2, epoch)
+        mlflow.log_metric('val_r2', val_r2, epoch)
+        mlflow.log_metric('val_loss', val_loss, epoch)
+        logging.info(f'val_loss type: {type(val_loss)}, val_len type: {type(val_len)}')
+        return val_loss, val_len, {"val_loss": val_loss, "train_loss": train_loss, "train_r2": train_r2, "val_r2": val_r2}
```

Comment:
this is a quite general question: why do we not inherit this and some other methods from the code for local models? as far as I understand, a FL client class should be very similar to a local model class. One of the main differences is to keep track of the round number in FL client but not in local. But many other things can be inherited I guess